Python 2.30 + TS 4.26 (#604)
dbos-inc/dbos-docs: docs/explanations/sharing-a-system-database.md

---
sidebar_position: 12
title: Sharing a System Database
---

Multiple DBOS applications, potentially in different languages, can share a single system database.
Each application is identified by its configured name and owns everything it creates: workflows, steps, queues, schedules, and application versions.
Applications sharing a system database are isolated from one another by default, but can freely interoperate by naming each other.
For example, one application can enqueue another's workflows and wait for their results.

## Application Names and Ownership

Every application is identified by the `name` in its configuration, so each application sharing a system database must have a distinct name.
Ownership determines which application runs what:

- A workflow is dequeued, run, and recovered only by the application that owns it.
- A queue is polled only by the application that registered it, even if another application enqueues workflows on it.

> - A schedule is fired only by the application that created it, and its workflows run on that application.

Reviewer: ```suggestion
- A schedule is fired only by the application that created it, and its workflows are owned by that application.
```

- Application versions are tracked per application, so one application's deployments do not affect which version its peers consider latest.

> Queue, schedule, and version names remain globally unique across all applications sharing a system database; registering a name that a different application already owns raises an error.

Reviewer: What happens if I manually lock-step my versions across applications?

Author: Not allowed, version names must be unique per application

Workflow IDs are also unique across the entire system database, so ID-addressed operations (retrieving a workflow's handle, status, or result by ID, and sending messages or reading events and streams) work across applications regardless of ownership.
Observability queries (`list_workflows`, `list_queues`, `list_schedules`) are scoped to the calling application by default.

## Calling Another Application's Workflows

To run another application's workflow, enqueue it by name, naming the application that implements it.
The enqueued workflow is owned by the target application, which dequeues and runs it on its latest application version.
Because workflow IDs are global, you can then wait for the result from the returned handle.

<Tabs groupId="language">
<TabItem value="python" label="Python">

```python

> from dbos import DBOS, EnqueueOptions

Reviewer: Suggest adding a comment

```python
# Application: order-client
```

or something like that

Author: Tried it, found introducing more terminology made things more confusing rather than less
